bqskit/passes/search/generators/discrete.py

```diff
@@ -0,0 +1,206 @@
+        except IndexError:
+            return False
+
+    def count_repeats(
+        self,
+        circuit: Circuit,
+        gate: Gate,
+        qudit: int,
+    ) -> int:
+        """Count the number of times the last gate is repeated on qudit."""
+        count = 0
+        for cycle in reversed(range(circuit.num_cycles)):
+            try:
+                op = circuit.get_operation((cycle, qudit))
+                if op.gate == gate:
+                    count += 1
+                else:
+                    return count
+            except IndexError:
+                continue
+        return count
+
+    def gen_successors(self, circuit: Circuit, data: PassData) -> list[Circuit]:
+        """
+        Generate the successors of a circuit node.
+
+        Raises:
+            ValueError: If circuit is a single-qudit circuit.
+        """
+        if not isinstance(circuit, Circuit):
+            raise TypeError(f'Expected circuit, got {type(circuit)}.')
+
+        if circuit.num_qudits < 2:
+            raise ValueError('Cannot expand a single-qudit circuit.')
+
+        # Get the coupling graph
+        coupling_graph = data.connectivity
+
+        # Generate successors
+        successors = []
+        hashes = set()
+        singles = [gate for gate in self.gateset if gate.num_qudits == 1]
+        multis = [gate for gate in self.gateset if gate.num_qudits > 1]
+
+        def add_to_successors(circuit: Circuit) -> None:
+            h = self.hash_circuit_structure(circuit)
+            if h not in hashes:
+                successors.append(circuit)
+                hashes.add(h)
+
+        for gate in singles:
+            for qudit in range(circuit.num_qudits):
+                if self.cancels_something(circuit, gate, (qudit,)):
+                    continue
+                if isinstance(gate, TGate):
+                    if self.count_repeats(circuit, TGate(), qudit) >= 7:
+                        continue
+                successor = circuit.copy()
+                successor.append_gate(gate, [qudit])
+
+                add_to_successors(successor)
+
+                if self.double_headed:
+                    successor = circuit.copy()
+                    op = Operation(gate, [qudit])
+                    successor.insert(0, op)
+                    add_to_successors(successor)
+
+        for gate in multis:
+            for edge in coupling_graph:
+                if self.cancels_something(circuit, gate, edge):
+                    continue
+                successor = circuit.copy()
+                successor.append_gate(gate, edge)
+                add_to_successors(successor)
+
+                if self.double_headed:
+                    successor = circuit.copy()
+                    op = Operation(gate, edge)
+                    successor.insert(0, op)
+                    add_to_successors(successor)
+
+        return successors
+
+    def hash_circuit_structure(self, circuit: Circuit) -> int:
```

Comment:
Hashing is expensive, and I only think it's necessary if `double_head` = True. You should only use it when necessary.

It also won't catch equivalence classes, which are probably more valuable for pruning the space, i.e, `ST == TS``

Reply:
When testing locally, using hashing when `double_headed = False` takes the same amount of time as `double_headed = True` (~15 seconds for a 2 qubit circuit with 4 gates and the default gate set). I think keeping hashing either way makes the pass simpler. 

Unfortunately there is not an efficient way to check equivalence classes for this. The `cancels_something` and `count_repeats` methods are essential for doing some amount of logical-equivalence pruning.